File: smart-parcel-backend/src/main/java/com/capstone/smart_parcel/dto/common/PageResponse.java
package com.capstone.smart_parcel.dto.common;

import org.springframework.data.domain.Page;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public record PageResponse<T>(
        List<T> content,
        long totalElements,
        int totalPages,
        int page,
        int size,
        boolean first,
        boolean last,
        String sort
) {

    public static <T> PageResponse<T> of(Page<T> page) {
        return of(page, Function.identity());
    }

    public static <T, R> PageResponse<R> of(Page<T> page, Function<T, R> mapper) {
        List<R> mapped = page.getContent().stream()
                .map(mapper)
                .collect(Collectors.toList());
        String sort = page.getSort().isSorted() ? page.getSort().toString() : "";
        return new PageResponse<>(
                mapped,
                page.getTotalElements(),
                page.getTotalPages(),
                page.getNumber(),
                page.getSize(),
                page.isFirst(),
                page.isLast(),
                sort
        );
    }
}

File: smart_parcel_mobile/lib/data/dto/page_response.dart
class PageResponse<T> {
  final List<T> content;
  final int totalElements;
  final int totalPages;
  final int page;
  final int size;
  final bool first;
  final bool last;
  final String? sort;

  const PageResponse({
    required this.content,
    required this.totalElements,
    required this.totalPages,
    required this.page,
    required this.size,
    required this.first,
    required this.last,
    this.sort,
  });

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) {
    final items = (json['content'] as List<dynamic>? ?? [])
        .map((item) => fromJsonT(item))
        .toList();
    return PageResponse(
      content: items,
      totalElements: (json['totalElements'] as num?)?.toInt() ?? 0,
      totalPages: (json['totalPages'] as num?)?.toInt() ?? 0,
      page: (json['page'] as num?)?.toInt() ?? 0,
      size: (json['size'] as num?)?.toInt() ?? items.length,
      first: json['first'] as bool? ?? true,
      last: json['last'] as bool? ?? true,
      sort: json['sort'] as String?,
    );
  }
}
